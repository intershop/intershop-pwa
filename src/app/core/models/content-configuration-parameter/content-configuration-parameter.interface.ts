export interface ContentConfigurationParameterData {
  definitionQualifiedName: string;
  value: string | object;
}
