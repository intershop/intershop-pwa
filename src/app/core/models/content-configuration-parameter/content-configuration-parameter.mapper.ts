import { Injectable } from '@angular/core';
import { Store, select } from '@ngrx/store';

import { getCurrentLocale, getICMStaticURL } from 'ish-core/store/core/configuration';
import { mapToProperty } from 'ish-core/utils/operators';

import { ContentConfigurationParameterData } from './content-configuration-parameter.interface';

export interface ContentConfigurationParameters {
  [key: string]: string | object | number;
}

@Injectable({ providedIn: 'root' })
export class ContentConfigurationParameterMapper {
  private staticURL: string;
  private lang = '-';

  constructor(store: Store) {
    store.pipe(select(getICMStaticURL)).subscribe(url => (this.staticURL = url));
    store.pipe(select(getCurrentLocale), mapToProperty('lang')).subscribe(lang => (this.lang = lang || '-'));
  }

  fromData(data: { [name: string]: ContentConfigurationParameterData }): ContentConfigurationParameters {
    let configurationParameters = {};

    if (data) {
      configurationParameters = Object.entries(data)
        .map(([key, value]) => ({ [key]: value.value }))
        .reduce((acc, val) => ({ ...acc, ...val }));
    }

    this.postProcessFileReferences(configurationParameters);

    return configurationParameters;
  }

  /**
   * TODO: make this dependant on the type of the configuration parameter once the CMS REST API provides this information
   * for now filter all configuration parameters that start with 'Image' or 'Video'
   * and where the value does not start with 'http' but contains ':/'
   + if the filter matches convert the CMS REST API value in a full server URL to the configured file
   */
  postProcessFileReferences(data: ContentConfigurationParameters): ContentConfigurationParameters {
    Object.keys(data)
      .filter(
        key =>
          (key.startsWith('Image') || key.startsWith('Video')) &&
          data[key] &&
          !data[key].toString().startsWith('http') &&
          data[key].toString().includes(':/')
      )
      .forEach(key => {
        const split = data[key].toString().split(':');
        data[key] = encodeURI(`${this.staticURL}/${split[0]}/${this.lang}${split[1]}`);
      });
    return data;
  }
}
