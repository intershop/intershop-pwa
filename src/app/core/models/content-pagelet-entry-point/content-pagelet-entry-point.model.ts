import { ContentConfigurationParameters } from 'ish-core/models/content-configuration-parameter/content-configuration-parameter.mapper';
import { SeoAttributes } from 'ish-core/models/seo-attributes/seo-attributes.model';

export interface ContentPageletEntryPoint {
  id: string;
  definitionQualifiedName: string;
  displayName: string;
  domain: string;
  resourceSetId: string;
  pageletIDs?: string[];
  configurationParameters?: ContentConfigurationParameters;
  seoAttributes?: SeoAttributes;
}
