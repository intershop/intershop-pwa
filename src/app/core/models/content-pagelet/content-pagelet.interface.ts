import { ContentConfigurationParameterData } from 'ish-core/models/content-configuration-parameter/content-configuration-parameter.interface';
import { ContentSlotData } from 'ish-core/models/content-slot/content-slot.interface';

export interface ContentPageletData {
  definitionQualifiedName: string;
  id: string;
  displayName: string;
  domain: string;
  configurationParameters?: { [name: string]: ContentConfigurationParameterData };
  slots?: { [definitionQualifiedName: string]: ContentSlotData };
}
