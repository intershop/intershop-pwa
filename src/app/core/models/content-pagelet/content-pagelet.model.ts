import { ContentConfigurationParameters } from 'ish-core/models/content-configuration-parameter/content-configuration-parameter.mapper';
import { ContentSlot } from 'ish-core/models/content-slot/content-slot.model';

export interface ContentPagelet {
  definitionQualifiedName: string;
  id: string;
  displayName: string;
  domain: string;
  configurationParameters?: ContentConfigurationParameters;
  slots?: ContentSlot[];
}
