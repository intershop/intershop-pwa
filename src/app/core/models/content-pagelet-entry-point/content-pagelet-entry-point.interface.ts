import { ContentConfigurationParameterData } from 'ish-core/models/content-configuration-parameter/content-configuration-parameter.interface';
import { ContentPageletData } from 'ish-core/models/content-pagelet/content-pagelet.interface';
import { Link } from 'ish-core/models/link/link.model';

export interface ContentPageletEntryPointData {
  id: string;
  definitionQualifiedName: string;
  displayName: string;
  domain: string;
  resourceSetId: string;
  pagelets: ContentPageletData[];
  configurationParameters?: { [name: string]: ContentConfigurationParameterData };
  link: Link;
}
