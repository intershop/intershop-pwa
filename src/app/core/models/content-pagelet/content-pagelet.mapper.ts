import { Injectable } from '@angular/core';

import { ContentConfigurationParameterMapper } from 'ish-core/models/content-configuration-parameter/content-configuration-parameter.mapper';
import { ContentSlotData } from 'ish-core/models/content-slot/content-slot.interface';
import { ContentSlot } from 'ish-core/models/content-slot/content-slot.model';

import { ContentPageletData } from './content-pagelet.interface';
import { ContentPagelet } from './content-pagelet.model';

@Injectable({ providedIn: 'root' })
export class ContentPageletMapper {
  constructor(private contentConfigurationParameterMapper: ContentConfigurationParameterMapper) {}

  fromData(data: ContentPageletData): ContentPagelet[] {
    if (!data) {
      throw new Error('falsy input');
    }

    const { definitionQualifiedName, id, domain, displayName } = data;
    const configurationParameters = this.contentConfigurationParameterMapper.fromData(data.configurationParameters);

    let slots: ContentSlot[] = [];
    let pagelets: ContentPagelet[] = [];
    if (data.slots) {
      const deep = Object.values(data.slots).map(x => this.fromSlotData(x));
      slots = deep.map(val => val.slot);
      pagelets = deep.map(val => val.pagelets).reduce((acc, val) => [...acc, ...val]);
    }

    return [
      {
        configurationParameters,
        definitionQualifiedName,
        id,
        slots,
        domain,
        displayName,
      },
      ...pagelets,
    ];
  }

  private fromSlotData(data: ContentSlotData): { slot: ContentSlot; pagelets: ContentPagelet[] } {
    if (!data) {
      throw new Error('falsy input');
    }

    const definitionQualifiedName = data.definitionQualifiedName;
    const configurationParameters = this.contentConfigurationParameterMapper.fromData(data.configurationParameters);
    const pageletIDs = !data.pagelets ? [] : data.pagelets.map(pagelet => pagelet.id);
    const displayName = data.displayName;

    const slot: ContentSlot = {
      definitionQualifiedName,
      configurationParameters,
      pageletIDs,
      displayName,
    };

    const pagelets = !data.pagelets
      ? []
      : data.pagelets.map(x => this.fromData(x)).reduce((acc, val) => [...acc, ...val], []);

    return { slot, pagelets };
  }
}
