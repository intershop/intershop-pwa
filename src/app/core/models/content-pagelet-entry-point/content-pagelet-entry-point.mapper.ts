import { Injectable } from '@angular/core';

import { ContentConfigurationParameterMapper } from 'ish-core/models/content-configuration-parameter/content-configuration-parameter.mapper';
import { ContentPageletEntryPointData } from 'ish-core/models/content-pagelet-entry-point/content-pagelet-entry-point.interface';
import { ContentPageletEntryPoint } from 'ish-core/models/content-pagelet-entry-point/content-pagelet-entry-point.model';
import { ContentPageletMapper } from 'ish-core/models/content-pagelet/content-pagelet.mapper';
import { ContentPagelet } from 'ish-core/models/content-pagelet/content-pagelet.model';

@Injectable({ providedIn: 'root' })
export class ContentPageletEntryPointMapper {
  constructor(
    private contentConfigurationParameterMapper: ContentConfigurationParameterMapper,
    private contentPageletMapper: ContentPageletMapper
  ) {}

  /**
   * Converts {@link ContentPageletEntryPointData} to the model entity {@link ContentPageletEntryPoint} and enclosed {@link ContentPagelet}s.
   */
  fromData(data: ContentPageletEntryPointData): [ContentPageletEntryPoint, ContentPagelet[]] {
    if (!data) {
      throw new Error('falsy input');
    }

    let pagelets: ContentPagelet[] = [];
    let pageletIDs: string[] = [];

    if (data.pagelets) {
      pageletIDs = data.pagelets.map(p => p.id);
      pagelets = data.pagelets
        .map(x => this.contentPageletMapper.fromData(x))
        .reduce((acc, val) => [...acc, ...val], []);
    }

    const configurationParameters = this.contentConfigurationParameterMapper.fromData(data.configurationParameters);

    const pageletEntryPoint: ContentPageletEntryPoint = {
      id: data.id,
      definitionQualifiedName: data.definitionQualifiedName,
      displayName: data.displayName,
      domain: data.domain,
      resourceSetId: data.resourceSetId,
      pageletIDs,
      configurationParameters,
    };

    return [pageletEntryPoint, pagelets];
  }
}
