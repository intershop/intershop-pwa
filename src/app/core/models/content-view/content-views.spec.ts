import { ContentPageletEntryPoint } from 'ish-core/models/content-pagelet-entry-point/content-pagelet-entry-point.model';
import { ContentConfigurationParameters } from '../content-configuration-parameter/content-configuration-parameter.mapper';
import { ContentPagelet } from '../content-pagelet/content-pagelet.model';

import {
  ContentConfigurationParameterView,
  ContentPageletEntryPointView,
  createContentConfigurationParameterView,
  createContentPageletEntryPointView,
} from './content-views';

describe('Content Views', () => {
  let configurationParameters: ContentConfigurationParameters;
  let pageletEntryPoint: ContentPageletEntryPoint;
  let pagelets: { [id: string]: ContentPagelet };

  beforeEach(() => {
    configurationParameters = {
      key1: '1',
      key2: 'true',
      key3: ['hello', 'world'],
      key4: { test: 'hello' },
    };

    pageletEntryPoint = {
      id: 'include',
      definitionQualifiedName: 'fq',
      domain: 'domain',
      resourceSetId: 'resId',
      displayName: 'name',
      pageletIDs: ['p1', 'p2'],
      configurationParameters,
    };

    pagelets = [
      {
        id: 'p1',
        domain: 'domain',
        displayName: 'p1',
        definitionQualifiedName: 'fq',
        configurationParameters: {
          key4: '2',
        },
        slots: [
          {
            definitionQualifiedName: 'fq',
            displayName: 'slot',
          },
        ],
      },
      {
        id: 'p2',
        displayName: 'p2',
        domain: 'domain',
        definitionQualifiedName: 'fq',
        slots: [
          {
            definitionQualifiedName: 'fq',
            displayName: 'slot',
            configurationParameters: {
              key5: 'test',
            },
            pageletIDs: ['p3'],
          },
        ],
      },
      {
        id: 'p3',
        displayName: 'p3',
        domain: 'domain',
        definitionQualifiedName: 'fq',
        slots: [
          {
            definitionQualifiedName: 'fq',
            displayName: 'slot',
            configurationParameters: {
              key6: '3',
            },
            pageletIDs: ['p4'],
          },
        ],
      },
      {
        id: 'p4',
        displayName: 'p4',
        domain: 'domain',
        definitionQualifiedName: 'fq',
        configurationParameters: {
          key7: '4',
        },
      },
    ]
      .map(pagelet => ({ [pagelet.id]: pagelet }))
      .reduce((acc, val) => ({ ...acc, ...val }));
  });

  it('should be able to create a view of a pagelet entry point', () => {
    expect(() => createContentPageletEntryPointView(pageletEntryPoint, pagelets)).not.toThrow();
    expect(createContentPageletEntryPointView(pageletEntryPoint, pagelets)).toMatchSnapshot();
  });

  it('should be able to create a view of configuration parameters', () => {
    expect(() => createContentConfigurationParameterView(configurationParameters)).not.toThrow();
    expect(createContentConfigurationParameterView(configurationParameters)).toMatchSnapshot();
  });

  describe('parameter view created', () => {
    let view: ContentConfigurationParameterView;

    beforeEach(() => {
      view = createContentConfigurationParameterView(configurationParameters);
    });

    describe('hasParam', () => {
      it('should return true for defined parameters', () => {
        expect(view.hasParam('key1')).toBeTrue();
        expect(view.hasParam('key2')).toBeTrue();
        expect(view.hasParam('key3')).toBeTrue();
        expect(view.hasParam('key4')).toBeTrue();
      });

      it('should return false for undefined parameters', () => {
        expect(view.hasParam('key5')).toBeFalse();
      });
    });

    describe('stringParam', () => {
      it('should return string values for parameters', () => {
        expect(view.stringParam('key1')).toEqual('1');
        expect(view.stringParam('key2')).toEqual('true');
        expect(view.stringParam('key3')).toEqual('["hello","world"]');
        expect(view.stringParam('key4')).toEqual('{"test":"hello"}');
      });

      it('should return string values for parameters even if default value is provided', () => {
        expect(view.stringParam('key1', 'test')).toEqual('1');
        expect(view.stringParam('key2', 'test')).toEqual('true');
        expect(view.stringParam('key3', 'test')).toEqual('["hello","world"]');
        expect(view.stringParam('key4', 'test')).toEqual('{"test":"hello"}');
      });

      it('should return undefined for undefined parameters', () => {
        expect(view.stringParam('key5')).toBeUndefined();
      });

      it('should return default value for undefined parameters if supplied', () => {
        expect(view.stringParam('key5', 'test')).toEqual('test');
      });
    });

    describe('booleanParam', () => {
      it('should return parsed boolean values for parameters', () => {
        expect(view.booleanParam('key1')).toBeFalse();
        expect(view.booleanParam('key2')).toBeTrue();
        expect(view.booleanParam('key3')).toBeFalse();
        expect(view.booleanParam('key4')).toBeFalse();
      });

      it('should return parsed boolean values for parameters even if default value is supplied', () => {
        expect(view.booleanParam('key1', true)).toBeFalse();
        expect(view.booleanParam('key2', false)).toBeTrue();
        expect(view.booleanParam('key3', true)).toBeFalse();
        expect(view.booleanParam('key4', true)).toBeFalse();
      });

      it('should return undefined for undefined parameters', () => {
        expect(view.booleanParam('key5')).toBeUndefined();
      });

      it('should return default value for undefined parameters if supplied', () => {
        expect(view.booleanParam('key5', true)).toBeTrue();
      });
    });

    describe('numberParam', () => {
      it('should return parsed number values for parameters or NaN', () => {
        expect(view.numberParam('key1')).toBe(1);
        expect(view.numberParam('key2')).toBeNaN();
        expect(view.numberParam('key3')).toBeNaN();
        expect(view.numberParam('key4')).toBeNaN();
      });

      it('should return parsed number values for parameters or default value on parsing error if it was provided', () => {
        expect(view.numberParam('key1', 5)).toBe(1);
        expect(view.numberParam('key2', 5)).toBe(5);
        expect(view.numberParam('key3', 5)).toBe(5);
        expect(view.numberParam('key4', 5)).toBe(5);
      });

      it('should return NaN for undefined parameters', () => {
        expect(view.numberParam('key5')).toBeNaN();
      });

      it('should return default value for undefined parameters if one was provided', () => {
        expect(view.numberParam('key5', 6)).toBe(6);
      });
    });

    describe('configParam', () => {
      it('should passthrough values for parameters with correct type', () => {
        expect(view.configParam('key1')).toEqual('1');
        expect(view.configParam('key2')).toEqual('true');
        expect(view.configParam<string[]>('key3')).toHaveLength(2);
        expect(view.configParam<string[]>('key3')).toEqual(['hello', 'world']);
        expect(view.configParam<{ test: string }>('key4')).toHaveProperty('test', 'hello');
        expect(view.configParam<{ test: string }>('key4').test).toEqual('hello');
        expect(view.configParam<{ test: string }>('key4')).toEqual({ test: 'hello' });
      });

      it('should return undefined for undefined parameters', () => {
        expect(view.configParam('key5')).toBeUndefined();
      });
    });
  });

  describe('include view created', () => {
    let view: ContentPageletEntryPointView;

    beforeEach(() => {
      view = createContentPageletEntryPointView(pageletEntryPoint, pagelets);
    });

    it('should have properties on first level', () => {
      expect(view.numberParam('key1')).toBe(1);
      expect(view.booleanParam('key2')).toBeTrue();
      expect(view.configParam<string[]>('key3')).toIncludeAllMembers(['hello', 'world']);
    });

    it('should have pagelets on top level', () => {
      expect(view.pagelets()).toHaveLength(2);
      expect(view.pagelets().map(p => p.id)).toIncludeAllMembers(['p1', 'p2']);
      expect(view.pagelets()[0]).not.toBeUndefined();
      expect(view.pagelets()[1]).not.toBeUndefined();
      expect(view.pagelets()).toMatchSnapshot();
    });

    it('should have a slot for p1', () => {
      expect(view.pagelets()[0]).toBeTruthy();
      const p1 = view.pagelets()[0];
      expect(p1.slot('fq')).toBeTruthy();
      const slot = p1.slot('fq');
      expect(slot.pagelets()).toBeEmpty();
      expect(slot.configParam('no')).toBeUndefined();
    });

    it('should have a slot view on p2', () => {
      expect(view.pagelets()[1].slot('fq')).toMatchSnapshot();
    });

    it('should have a pagelet on slot view on p2', () => {
      expect(
        view
          .pagelets()[1]
          .slot('fq')
          .pagelets()
      ).toHaveLength(1);
      expect(
        view
          .pagelets()[1]
          .slot('fq')
          .pagelets()
      ).toMatchSnapshot();
    });

    it('should have deepest level navigateable on tree', () => {
      const p4 = view
        .pagelets()[1]
        .slot('fq')
        .pagelets()[0]
        .slot('fq')
        .pagelets()[0];

      expect(p4.slot('fq')).toBeUndefined();
      expect(p4.hasParam('key7')).toBeTrue();
      expect(p4.numberParam('key7')).toBe(4);
      expect(p4).toMatchSnapshot();
    });
  });
});
