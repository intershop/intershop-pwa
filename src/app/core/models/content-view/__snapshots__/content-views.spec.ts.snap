// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content Views include view created should have a pagelet on slot view on p2 1`] = `
Array [
  Object {
    "booleanParam": [Function],
    "configParam": [Function],
    "configurationParameters": undefined,
    "definitionQualifiedName": "fq",
    "displayName": "p3",
    "domain": "domain",
    "hasParam": [Function],
    "id": "p3",
    "numberParam": [Function],
    "slot": [Function],
    "stringParam": [Function],
  },
]
`;

exports[`Content Views include view created should have a slot view on p2 1`] = `
Object {
  "booleanParam": [Function],
  "configParam": [Function],
  "displayName": "slot",
  "hasParam": [Function],
  "numberParam": [Function],
  "pagelets": [Function],
  "stringParam": [Function],
}
`;

exports[`Content Views include view created should have deepest level navigateable on tree 1`] = `
Object {
  "booleanParam": [Function],
  "configParam": [Function],
  "configurationParameters": Object {
    "key7": "4",
  },
  "definitionQualifiedName": "fq",
  "displayName": "p4",
  "domain": "domain",
  "hasParam": [Function],
  "id": "p4",
  "numberParam": [Function],
  "slot": [Function],
  "stringParam": [Function],
}
`;

exports[`Content Views include view created should have pagelets on top level 1`] = `
Array [
  Object {
    "booleanParam": [Function],
    "configParam": [Function],
    "configurationParameters": Object {
      "key4": "2",
    },
    "definitionQualifiedName": "fq",
    "displayName": "p1",
    "domain": "domain",
    "hasParam": [Function],
    "id": "p1",
    "numberParam": [Function],
    "slot": [Function],
    "stringParam": [Function],
  },
  Object {
    "booleanParam": [Function],
    "configParam": [Function],
    "configurationParameters": undefined,
    "definitionQualifiedName": "fq",
    "displayName": "p2",
    "domain": "domain",
    "hasParam": [Function],
    "id": "p2",
    "numberParam": [Function],
    "slot": [Function],
    "stringParam": [Function],
  },
]
`;

exports[`Content Views should be able to create a view of a pagelet entry point 1`] = `
Object {
  "booleanParam": [Function],
  "configParam": [Function],
  "displayName": "name",
  "domain": "domain",
  "hasParam": [Function],
  "id": "include",
  "numberParam": [Function],
  "pagelets": [Function],
  "resourceSetId": "resId",
  "stringParam": [Function],
}
`;

exports[`Content Views should be able to create a view of configuration parameters 1`] = `
Object {
  "booleanParam": [Function],
  "configParam": [Function],
  "hasParam": [Function],
  "numberParam": [Function],
  "stringParam": [Function],
}
`;
